### PR TITLE
Add the fuzzer from oss-fuzz and a test that exercises it

### PR DESCRIPTION
**Pull request by derwolfe**
As part of the steps towards an [ideal-integration](https://github.com/google/oss-fuzz/blob/master/docs/ideal_integration.md), I'd like to move the fuzzer out of OSS-fuzz and have it built against the library as part of CI (this would apply to any new fuzzers as well). The goal of this work is to ensure that the fuzzer is always able to build and does not bit rot.

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -37,6 +37,7 @@ LIST (APPEND check_PROGRAMS
     version.cpp
     visitor.cpp
     zone.cpp
+
```
Comment (derwolfe) [inline test/CMakeLists.txt]: remove this newline.

**Comment by derwolfe** [comment]
This is the fuzzer currently being used by oss-fuzz (https://github.com/google/oss-fuzz/blob/e01619962ed33bce27e0d8a6b2c04e624967d531/projects/msgpack-c/unpack_pack_fuzzer.cc)

**Comment by redboltz** [comment]
@derwolfe , thank you for sending the PR.
I'm still learning how OSS-Fuzz works.

I tried to get the source code and input data (if exists) that reproduces the issue. So I downloaded msgpack-c%2Fmsgpack-c-memory-201804131609.zip from https://oss-fuzz.com/v2/testcase-detail/6022481354686464?noredirect=1. And unzip it.
But I couldn't find a source code. 
It seems that `unpack_pack_fuzzer` is the binary to reproduce the problem. 
I downloaded minimized test case and unminimized test case. 
Then I passed them to `unpack_pack_fuzzer`, the problem seems to be reproduced.
But without the test source code, I stuck.

Now, I understand that the following code is the source code:
https://github.com/google/oss-fuzz/blob/e01619962ed33bce27e0d8a6b2c04e624967d531/projects/msgpack-c/unpack_pack_fuzzer.cc

Then OSS-Fuzz read minimized test case or unminimized test case, and pass it as `data` and `size`.
```cpp
 int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
```

And I also checked https://github.com/google/oss-fuzz/tree/e01619962ed33bce27e0d8a6b2c04e624967d531/projects/msgpack-c files. Then I understand the container based environment.

I don't understand [infra](https://github.com/google/oss-fuzz/tree/e01619962ed33bce27e0d8a6b2c04e624967d531/infra) part yet. But I can restart my work :)

Let me clarify the point of the PR. The PR is the replacement of https://github.com/google/oss-fuzz/blob/e01619962ed33bce27e0d8a6b2c04e624967d531/projects/msgpack-c/unpack_pack_fuzzer.cc?
The advantage of the replacement is make sure `int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)` works if `msgpack-c` is updated. Am I understanding correctly?

**Comment by derwolfe** [comment]
> Let me clarify the point of the PR. The PR is the replacement of https://github.com/google/oss-fuzz/blob/e01619962ed33bce27e0d8a6b2c04e624967d531/projects/msgpack-c/unpack_pack_fuzzer.cc? The advantage of the replacement is make sure int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) works if msgpack-c is updated. Am I understanding correctly?

Yes! That's right. This will make it so that the fuzzer is less likely to break. Also, it has the added benefit of making it easier for you and other contributors to add new fuzzers to the project. Once/if this PR is merged, I'll start simplifying the logic that lives in oss-fuzz  to only be scripts necessary to build the fuzzers for oss-fuzz.

> I tried to get the source code and input data (if exists) that reproduces the issue. So I downloaded msgpack-c%2Fmsgpack-c-memory-201804131609.zip from https://oss-fuzz.com/v2/testcase-detail/6022481354686464?noredirect=1. And unzip it.
But I couldn't find a source code.
It seems that unpack_pack_fuzzer is the binary to reproduce the problem.
I downloaded minimized test case and unminimized test case.
Then I passed them to unpack_pack_fuzzer, the problem seems to be reproduced.
But without the test source code, I stuck. 
Now, I understand that the following code is the source code:
https://github.com/google/oss-fuzz/blob/e01619962ed33bce27e0d8a6b2c04e624967d531/projects/msgpack-c/unpack_pack_fuzzer.cc

It is likely easiest to reproduce the crasher using the dockerized infrastructure that they provide - as it sounds like you've done. The way I've typically worked with debugging such a bug is to use their containerized infrastructure and GDB. They have a container with GDB prebuilt and instructions for using it at https://github.com/google/oss-fuzz/blob/60835ac4207d5877f119129a2a1e8639ee4b92a3/docs/debugging.md. 

I've normally run debug sessions for reproducers using:

1. `git clone https://github.com/google/oss-fuzz.git`
2. download the reproducer and place it in `/build/out/msgpack-c/`, e.g. `/build/out/msgpack-c/the_name_of_the_reproducer`.
3. `docker run -ti --privileged -v $(pwd)/build/out/msgpack-c:/out -w /out gcr.io/oss-fuzz-base/base-runner-debug`. This will give you a shell in the container and you'll be able to run gdb
4.  In the container use gdb like usual. `gdb unpack_pack_fuzzer` then `r the_name_of_the_reproducer`

Thank you :)!

**Comment by redboltz** [comment]
Thank you! I merged the PR.

**Pull request merged**